src/System.Net.Security/src/System/Net/SecureProtocols/SslStreamInternal.cs

```diff
@@ -423,15 +588,31 @@ private int StartFrameBody(int readBytes, byte[] buffer, int offset, int count)
 
             EnsureInternalBufferSize(SecureChannel.ReadHeaderSize, readBytes);
 
-            readBytes = _Reader.ReadPacket(InternalBuffer, SecureChannel.ReadHeaderSize, readBytes);
+            if (asyncRequest != null)
+            {
+                asyncRequest.SetNextRequest(InternalBuffer, SecureChannel.ReadHeaderSize, readBytes, _readFrameCallback);
+
+                _reader.AsyncReadPacket(asyncRequest);
 
-            return ProcessFrameBody(readBytes, buffer, offset, count);
+                if (!asyncRequest.MustCompleteSynchronously)
+                {
+                    return 0;
+                }
+                
+                readBytes = asyncRequest.Result;
+            }
+            else
+            {
+                readBytes = _reader.ReadPacket(InternalBuffer, SecureChannel.ReadHeaderSize, readBytes);
+            }
+            
+            return ProcessFrameBody(readBytes, buffer, offset, count, asyncRequest);
         }
 
         //
         // readBytes == SSL Data Payload size on input or 0 on EOF
         //
-        private int ProcessFrameBody(int readBytes, byte[] buffer, int offset, int count)
```

Comment:
nit:  comments style is different from our standard.  Usually this would be written (with a period) as:

```c#
// readBytes == SSL Data Payload size on input or 0 on EOF.
```

check rest of file.